### PR TITLE
Encryption with DPAPI_NG CERTIFICATE

### PR DESCRIPTION
**Pull request by danleydmello**

### CODE REVIEW

File: mapcrypto.c [hunk 1]
```diff
@@ -673,6 +684,199 @@ void msDecryptStringWithDPAPI( const char * in, char * out )
...
+   if( thumbprint == NULL )
+   {
+      msSetError( MS_MISCERR,
+         "Require certifcate thumbprint to encrypt the data. The certificate must be stored in the Current user stored",
+         "msEncryptStringWithDPAPING()" );
+      return MS_FAILURE;
+   }
+
+   //Set the protection descriptor string with the thumbprint;
+   descriptor = (char*)msSmallMalloc( sizeof( char ) * (strlen( thumbprint ) + strlen( prefix )) );
```
Comment (simen-wu) [inline mapcrypto.c]: I would like to add an extra space for the trailing 0 here.

File: mapcrypto.c [hunk 2]
```diff
@@ -673,6 +684,199 @@ void msDecryptStringWithDPAPI( const char * in, char * out )
...
+   strcpy( descriptor, prefix );
+   strcat( descriptor, thumbprint );
+
+   //Convert the protection descriptor to wide string because NCryptCreateProtectionDescriptor takes wide string.
+   size_t wn = mbsrtowcs( NULL, &descriptor, 0, NULL );
+   wchar_t * pwdescriptor = (wchar_t *)msSmallMalloc( sizeof( wchar_t ) * strlen( descriptor ) );
```
Comment (simen-wu) [inline mapcrypto.c]: Again, alloc 1 more space for the trailing '\0' just in case.

File: mapcrypto.c [hunk 3]
```diff
@@ -673,6 +684,199 @@ void msDecryptStringWithDPAPI( const char * in, char * out )
...
+      return MS_FAILURE;
+   }
+
+   //Set the protection descriptor string with the thumbprint;
+   descriptor = (char*)msSmallMalloc( sizeof( char ) * (strlen( thumbprint ) + strlen( prefix )) );
+   strcpy( descriptor, prefix );
+   strcat( descriptor, thumbprint );
+
+   //Convert the protection descriptor to wide string because NCryptCreateProtectionDescriptor takes wide string.
+   size_t wn = mbsrtowcs( NULL, &descriptor, 0, NULL );
```
Comment (simen-wu) [inline mapcrypto.c:715]: I'm not quite sure parameter 0 here. Does it mean it's going to check all the text in descriptor?

**Review by simen-wu — APPROVED** [review]

File: mapcrypto.c (hunk 2 above)
Comment (mdastous-bentley) [reply mapcrypto.c]: Where is pwdescriptor freed?

File: mapcrypto.c (hunk 1 above)
Comment (mdastous-bentley) [reply mapcrypto.c]: Where is descriptor freed?

Comment (danleydmello) [reply mapcrypto.c]: it gets freed by mbsrtowcs . but I will freed it up anyways..

File: mapcrypto.c (hunk 2 above)
Comment (danleydmello) [reply mapcrypto.c]: Ooops, I missed that somehow. Thanks

File: mapcrypto.c (hunk 3 above)
Comment (danleydmello) [reply mapcrypto.c:715]: **It  indicate the number of wide characters available in the array pointed to by destination pointer, In this call, this is what we are finding by passing NULL dst pointer.
--**

**Pull request merged**